File: examples/chp-develops-chp/agent.py
#!/usr/bin/env python3
"""CHP-develops-CHP — a self-hosted coding/analysis agent running entirely on the mesh.

The reasoning model (Qwen3 via chp.adapters.mlx on the inference node) drives a
ReAct loop whose *tools are CHP capabilities*, routed across the fleet through the
gateway and recorded as evidence:

  reason (mlx.chat @ inference)
    └─ act  → scout / filesystem / conformance  (@ primary, repo node)
         └─ observe → feed result back → repeat (bounded)

Every tool call is a governed capability invocation sharing one correlation id, so
the whole run is replayable (`chp-host` mesh tooling / router.replay). Mutating
actions are assessed by chp.adapters.safety first and only run under --apply.

Usage:
    export CHP_GATEWAY_KEY=$(chp-host secrets get CHP_HOST_API_KEY)
    python agent.py "Summarise what the mlx adapter does and suggest one improvement"
    python agent.py --repo /path/to/chp-core --steps 6 "your task"

This is a demonstration of the agentic layer (see docs/agentic-mesh.md), not an
unattended committer: by default it is read-only and proposes; it never pushes.
"""

from __future__ import annotations

import argparse
import json
import os
import re
import sys
import uuid

from chp_core.http import RemoteCapabilityHost

DEFAULT_GATEWAY = "http://127.0.0.1:8800"
DEFAULT_REPO = "/Users/patrickschmied/Projects/capabilityhostprotocol/chp-core"
DEFAULT_MODEL = "mlx-community/Qwen3-4B-Instruct-2507-4bit"

# Read-only tools the agent may call. Each maps to a CHP capability + the node it
# must run on (prefer), plus how to render the payload from the model's args.
TOOLS: dict[str, dict] = {
    "list_dir": {
        "cap": "chp.adapters.filesystem.list_directory", "prefer": "primary",
        "args": ["path"], "help": "List a directory. args: {path}"},
    "read_file": {
        "cap": "chp.adapters.filesystem.read_file", "prefer": "primary",
        "args": ["path"], "help": "Read a file. args: {path}"},
    "grep": {
        "cap": "chp.adapters.filesystem.grep", "prefer": "primary",
        "args": ["pattern", "path"], "help": "Regex search. args: {pattern, path}"},
    "scout": {
        "cap": "chp.adapters.scout.query", "prefer": "primary",
        "args": ["task"], "help": "Find relevant code for a task. args: {task}"},
    "conformance": {
        "cap": "chp.adapters.conformance.check_source", "prefer": "primary",
        "args": ["source_path"], "help": "Check a source file for CHP violations. args: {source_path}"},
}

SYSTEM_PROMPT = """You are a coding analyst working INSIDE the Capability Host Protocol mesh.
You answer the user's task by calling tools that read the CHP repository, then giving a final answer.

You MUST reply with exactly ONE JSON object per turn, nothing else:
  - To call a tool:   {{"tool": "<name>", "args": {{...}}}}
  - To finish:        {{"final": "<your answer, citing files you read>"}}

Available tools:
{tools}

Rules:
- The repository root is: {repo}
- Use absolute paths under the repository root.
- Call tools to gather evidence BEFORE answering. Prefer scout/grep to locate, then read_file.
- Keep to at most {steps} tool calls, then give {{"final": ...}}.
- Output ONLY the JSON object. No prose, no markdown fences."""


def _extract_json(text: str) -> dict | None:
    """Lenient: find the first balanced {...} object in the model's reply."""
    start = text.find("{")
    while start != -1:
        depth = 0
        for i in range(start, len(text)):
            if text[i] == "{":
                depth += 1
            elif text[i] == "}":
                depth -= 1
                if depth == 0:
                    try:
                        return json.loads(text[start:i + 1])
                    except json.JSONDecodeError:
                        break
        start = text.find("{", start + 1)
    return None


class CHPDevAgent:
    def __init__(self, gateway: str, key: str, repo: str, model: str, steps: int):
        self.gw = RemoteCapabilityHost(gateway, api_key=key, timeout=180)
        self.repo, self.model, self.max_steps = repo, model, steps
        self.correlation = {"correlation_id": f"chp-dev-{uuid.uuid4().hex[:8]}"}
        self.trail: list[dict] = []

    def _invoke(self, cap: str, payload: dict, prefer: str | None = None) -> tuple[bool, object]:
        meta = {"prefer": prefer} if prefer else {}
        r = self.gw.invoke(cap, payload, correlation=self.correlation, metadata=meta)
        self.trail.append({"cap": cap, "outcome": r.outcome, "prefer": prefer})
        return r.success, (r.data if r.success else (r.error or r.denial))

    def _chat(self, messages: list[dict]) -> str:
        ok, data = self._invoke(
            "chp.adapters.mlx.chat",
            {"model": self.model, "messages": messages, "max_tokens": 700, "temperature": 0.2},
            prefer="inference")
        if not ok:
            raise RuntimeError(f"mlx.chat failed: {data}")
        return (data.get("message") or {}).get("content", "")

    def _run_tool(self, name: str, args: dict) -> str:
        spec = TOOLS.get(name)
        if not spec:
            return f"ERROR: unknown tool {name!r}. Valid: {', '.join(TOOLS)}"
        payload = {k: args.get(k) for k in spec["args"] if args.get(k) is not None}
        if name == "scout":
            payload["repo_path"] = self.repo
        ok, data = self._invoke(spec["cap"], payload, prefer=spec["prefer"])
        if not ok:
            return f"ERROR: {json.dumps(data)[:300]}"
        # Compact the observation so the small model isn't overwhelmed.
        return json.dumps(data)[:1500]

    def run(self, task: str) -> str:
        tools_doc = "\n".join(f"  - {n}: {s['help']}" for n, s in TOOLS.items())
        system = SYSTEM_PROMPT.format(tools=tools_doc, repo=self.repo, steps=self.max_steps)
        messages = [{"role": "system", "content": system},
                    {"role": "user", "content": task}]

        for step in range(1, self.max_steps + 1):
            reply = self._chat(messages)
            action = _extract_json(reply)
            if not action:
                messages.append({"role": "assistant", "content": reply})
                messages.append({"role": "user", "content":
                                 'Reply with ONE JSON object: {"tool":...} or {"final":...}.'})
                print(f"  [{step}] (no JSON, nudging)")
                continue
            if "final" in action:
                print(f"  [{step}] final answer")
                return str(action["final"])
            tool, args = action.get("tool"), action.get("args", {}) or {}
            print(f"  [{step}] tool: {tool}({json.dumps(args)[:80]})")
            obs = self._run_tool(tool, args)
            messages.append({"role": "assistant", "content": json.dumps(action)})
            messages.append({"role": "user", "content": f"Observation: {obs}"})

        # Out of steps — ask for a final answer from what was gathered.
        messages.append({"role": "user", "content": 'Step budget reached. Give {"final": ...} now.'})
        final = _extract_json(self._chat(messages)) or {}
        return str(final.get("final", "(no final answer produced)"))


def main() -> int:
    ap = argparse.ArgumentParser(description="CHP-develops-CHP agent (runs on the mesh).")
    ap.add_argument("task", help="What the agent should investigate/do.")
    ap.add_argument("--gateway", default=os.environ.get("CHP_GATEWAY", DEFAULT_GATEWAY))
    ap.add_argument("--repo", default=DEFAULT_REPO, help="Repository root on the primary node.")
    ap.add_argument("--model", default=os.environ.get("MLX_MODEL", DEFAULT_MODEL))
    ap.add_argument("--steps", type=int, default=6, help="Max tool calls.")
    args = ap.parse_args()

    key = os.environ.get("CHP_GATEWAY_KEY") or os.environ.get("CHP_HOST_API_KEY")
    if not key:
        print("Set CHP_GATEWAY_KEY (e.g. export CHP_GATEWAY_KEY=$(chp-host secrets get CHP_HOST_API_KEY))",
              file=sys.stderr)
        return 2

    agent = CHPDevAgent(args.gateway, key, args.repo, args.model, args.steps)
    print(f"CHP-develops-CHP  ·  correlation={agent.correlation['correlation_id']}  ·  model={args.model}")
    print(f"task: {args.task}\n")
    answer = agent.run(args.task)

    print("\n" + "=" * 70 + "\nANSWER\n" + "=" * 70)
    print(answer)
    print("\n" + "-" * 70)
    print(f"evidence: {len(agent.trail)} governed capability calls under correlation "
          f"{agent.correlation['correlation_id']}")
    for t in agent.trail:
        print(f"  {t['outcome']:8} {t['cap']}" + (f"  (@{t['prefer']})" if t['prefer'] else ""))
    print(f"\nReplay the full federated trail with this correlation id via the mesh evidence tooling.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: harness/src/agent.ts
/**
 * CHP × OpenHarness — Phase 0 spike.
 *
 * An OpenHarness Agent whose tools ARE the CHP mesh: it connects (MCP, stdio) to
 * `chp-host mcp --environment <mesh>`, so every governed capability across the
 * fleet (filesystem, scout, conformance, git, host, mlx, …) becomes a tool. The
 * model is a local MLX server (OpenAI-compatible) on the primary.
 *
 * Each MCP tool call carries the chp-host MCP-session correlation id, so the run
 * is replayable via the CHP evidence tooling (`chp-host mesh audit` / replay).
 *
 *   # terminal A is implicit — the stdio transport spawns chp-host mcp for us
 *   export CHP_MODEL_BASE_URL=http://localhost:8081/v1   # primary Qwen3-14B
 *   export CHP_MODEL_ID=mlx-community/Qwen3-14B-4bit
 *   npm run agent -- "What capabilities does the mlx adapter expose? Cite the file."
 */

import { homedir } from "node:os";
import { createOpenAICompatible } from "@ai-sdk/openai-compatible";
import { randomUUID } from "node:crypto";
import { appendFileSync, mkdirSync } from "node:fs";
import { dirname } from "node:path";
import chalk from "chalk";
import {
  Agent,
  Conversation,
  toRunner,
  apply,
  withTurnTracking,
  withRetry,
  withPersistence,
  connectMCPServers,
  closeMCPClients,
  type ToolCallInfo,
  type SessionStore,
} from "@openharness/core";

// ── Config ──────────────────────────────────────────────────────────────
const MESH = process.env.CHP_MESH ?? `${homedir()}/.chp/mesh.json`;
// Default to the gateway's OpenAI shim → inference runs as a capacity-routed,
// evidenced mesh capability (chp.adapters.mlx.chat). Override to hit a node directly.
const MODEL_BASE_URL = process.env.CHP_MODEL_BASE_URL ?? "http://127.0.0.1:8800/v1";
const MODEL_ID = process.env.CHP_MODEL_ID ?? "mlx-community/Qwen3-4B-Instruct-2507-4bit";
const GATEWAY = process.env.CHP_GATEWAY ?? "http://127.0.0.1:8800";
const GATEWAY_KEY = process.env.CHP_GATEWAY_KEY ?? process.env.CHP_HOST_API_KEY ?? "";
// The safety adapter lives on the primary host (the gateway doesn't surface it).
const SAFETY_URL = process.env.CHP_SAFETY_URL ?? "http://127.0.0.1:8803";
// require_approval verdicts: deny by default (safe); set CHP_AUTO_APPROVE=1 to allow.
const AUTO_APPROVE = process.env.CHP_AUTO_APPROVE === "1";

// ── Governance: assess a capability via chp.adapters.safety.assess ───────
async function hostInvoke(base: string, capabilityId: string, payload: unknown): Promise<any> {
  const res = await fetch(`${base}/invoke`, {
    method: "POST",
    headers: { "Content-Type": "application/json", "X-CHP-Key": GATEWAY_KEY },
    body: JSON.stringify({ capability_id: capabilityId, payload }),
  });
  return res.json();
}

async function assess(capId: string, input: unknown): Promise<{ level: string; recommendation: string }> {
  try {
    const r = await hostInvoke(SAFETY_URL, "chp.adapters.safety.assess", { capability_id: capId, payload: input });
    const d = r?.data ?? {};
    if (r?.outcome !== "success") throw new Error(JSON.stringify(r?.error ?? r));
    return { level: d.level ?? "unknown", recommendation: d.recommendation ?? "allow" };
  } catch (e) {
    // Fail CLOSED for clearly mutating tools, open otherwise — never silently allow a write we couldn't assess.
    const risky = /write|delete|update|restart|stop|install|push|exec|bash|run|create/i.test(capId);
    console.warn(`    ⚠ safety.assess unavailable (${String(e).slice(0, 80)}) — ${risky ? "denying (fail-closed)" : "allowing low-risk"}`);
    return { level: "unknown", recommendation: risky ? "block" : "allow" };
  }
}
// Node keys the mesh router needs to authenticate to remote hosts (names only).
const KEYCHAIN_KEYS = [
  "CHP_HOST_API_KEY",
  "CHP_PEER_0_KEY",
  "CHP_PEER_1_KEY",
  "CHP_NAS_KEY",
  "CHP_NAS_IP",
];

const SYSTEM_PROMPT =
  "You are an agent operating over the Capability Host Protocol (CHP) mesh. " +
  "Your tools are governed CHP capabilities spanning a fleet of nodes (filesystem, " +
  "code search/scout, conformance, git, host stats, and local LLM inference). " +
  "Use them to gather evidence and answer the user's task; cite the files/paths you read. " +
  "Prefer read-only tools first. Be concise.";

// MCP tool name → CHP capability id (mcp_server uses cap_id.replace(".", "_")).
const capMap = new Map<string, string>();
async function buildCapMap(): Promise<void> {
  const res = await fetch(`${GATEWAY}/capabilities`, { headers: { "X-CHP-Key": GATEWAY_KEY } });
  const data = await res.json();
  for (const c of data.capabilities ?? []) capMap.set(String(c.id).replaceAll(".", "_"), c.id);
}

// P1 governance bridge: assess every tool call via chp.adapters.safety.assess.
async function approve(toolCall: ToolCallInfo): Promise<boolean> {
  const capId = capMap.get(toolCall.toolName) ?? toolCall.toolName;
  const { level, recommendation } = await assess(capId, toolCall.input);
  const tag = recommendation === "block" ? chalk.red("BLOCK")
    : recommendation === "require_approval" ? chalk.yellow("APPROVE?")
    : chalk.green("allow");
  console.log(`  ${chalk.yellow("○")} ${chalk.cyan(toolCall.toolName)} ` +
    `${chalk.dim(`[safety:${level} → ${recommendation}]`)} ${tag}`);
  if (recommendation === "block") { console.log(`    ${chalk.red("✗ denied by safety.assess")}`); return false; }
  if (recommendation === "require_approval" && !AUTO_APPROVE) {
    console.log(`    ${chalk.red("✗ requires approval (set CHP_AUTO_APPROVE=1 to allow)")}`); return false;
  }
  return true;
}

// A1 — small models misread structured tool output (scout returned `files`, the 4B
// still said "not found"). Flatten the common shapes into explicit text before the
// model sees them, and wrap each tool's execute to apply it.
function summarizeToolResult(raw: any): any {
  let data: any = raw;
  if (typeof raw === "string") { try { data = JSON.parse(raw); } catch { return raw; } }
  // MCP results often wrap the payload in a content array of {type:'text', text}.
  if (data && Array.isArray(data.content)) {
    const t = data.content.find((c: any) => c?.type === "text")?.text;
    if (t) { try { data = JSON.parse(t); } catch { return t; } }
  }
  const obj = data && typeof data === "object" ? data : {};
  const files = obj.files ?? obj.citations;
  if (Array.isArray(files) && files.length) {
    const lines = files.map((f: any) =>
      typeof f === "string" ? f : `${f.path}${f.line_range ? ":" + f.line_range : ""}`);
    return `Found ${files.length} file(s):\n${lines.map((l: string) => "- " + l).join("\n")}\n` +
      "(Report these path(s)/line-range(s) as the answer — do NOT say 'not found'.)";
  }
  if (Array.isArray(obj.violations)) {
    return obj.violations.length
      ? `Conformance: ${obj.violations.length} violation(s):\n` +
        obj.violations.map((v: any) => "- " + JSON.stringify(v)).join("\n")
      : "Conformance: clean (no violations).";
  }
  return raw;
}

function wrapToolResults(tools: Record<string, any>): Record<string, any> {
  return Object.fromEntries(Object.entries(tools).map(([n, t]) => {
    const orig = t?.execute;
    if (typeof orig !== "function") return [n, t];
    return [n, { ...t, execute: async (args: any, opts: any) => summarizeToolResult(await orig(args, opts)) }];
  }));
}

async function main() {
  const task = process.argv.slice(2).join(" ").trim();
  if (!task) {
    console.error('Usage: npm run agent -- "your task"');
    process.exit(2);
  }

  console.log(chalk.bold.cyan("chp-harness") + chalk.dim(`  · model ${MODEL_ID} · mesh ${MESH}`));
  console.log(chalk.dim("connecting to the mesh over MCP (chp-host mcp)..."));

  const { clients, tools: allTools } = await connectMCPServers({
    chp: {
      type: "stdio",
      command: "chp-host",
      args: ["mcp", "--environment", MESH, "--secrets-from-keychain", ...KEYCHAIN_KEYS],
    },
  });
  // The mesh exposes ~150 capabilities; a small local model can't reason over that
  // many tool schemas. Curate to a task-relevant subset (override via CHP_TOOLS,
  // comma-separated substrings; empty = all).
  const filter = (process.env.CHP_TOOLS ?? "filesystem,conformance,host_stats,host_version,scout")
    .split(",").map((s) => s.trim()).filter(Boolean);
  const curated = filter.length
    ? Object.fromEntries(Object.entries(allTools).filter(([n]) => filter.some((f) => n.includes(f))))
    : allTools;
  const tools = wrapToolResults(curated); // A1: format structured results for small models
  console.log(chalk.dim(`mesh tools: ${Object.keys(allTools).length} available → ${Object.keys(tools).length} enabled`));

  // P1 governance: build the tool→capability map so every tool call can be assessed.
  await buildCapMap();
  console.log(chalk.dim(`governance: every tool call assessed via chp.adapters.safety.assess`));

  // X-CHP-Key authenticates the gateway shim; harmless when pointing at a node directly.
  const model = createOpenAICompatible({
    name: "mlx", baseURL: MODEL_BASE_URL, headers: { "X-CHP-Key": GATEWAY_KEY },
  }).chatModel(MODEL_ID);

  // P3 — a read-only `explore` subagent (mesh read tools). The main agent delegates
  // exploration to it via the auto-injected `task` tool; it keeps repo-spelunking
  // out of the main context (OpenHarness runs subagents with their own window).
  const readOnly = Object.fromEntries(
    Object.entries(tools).filter(([n]) => /read_file|list_directory|glob|grep|scout|stat_path/.test(n)),
  );
  const explore = new Agent({
    name: "explore",
    description: "Read-only mesh exploration — search and read repo files across the fleet.",
    systemPrompt:
      "You explore the CHP repo over the mesh (read-only). Call scout_query (or grep) to locate code. " +
      "Scout returns a `files` list of {path, line_range} — report each one VERBATIM as your finding. " +
      "NEVER answer 'not found' if scout returned any files; pass the paths/line ranges through.",
    model,
    tools: readOnly,
    maxSteps: 12,
  });
  console.log(chalk.dim(`subagent: explore (${Object.keys(readOnly).length} read-only tools)`));

  const agent = new Agent({
    name: "chp-agent",
    systemPrompt: SYSTEM_PROMPT +
      " For read-only exploration (finding/reading code), delegate to the `explore` subagent via the task tool.",
    model,
    tools,
    maxSteps: 12,
    approve,
    subagents: [explore],
    onSubagentEvent: (path: string[], ev: any) => {
      if (ev.type === "tool.done") console.log(chalk.magenta(`    ⌁ ${path.join(">")}: ${ev.toolName}`));
      if (ev.type === "done") console.log(chalk.magenta(`    ⌁ ${path.join(">")} done`));
    },
  });

  // C1 — opt-in transcript capture (CHP_CAPTURE_TRACES) feeds the flywheel corpus.
  const sessionId = randomUUID();
  const messageStore = new Map<string, any[]>();
  const store: SessionStore = {
    async load(id) { return messageStore.get(id); },
    async save(id, m) { messageStore.set(id, m); },
  };
  const runner = apply(toRunner(agent), withTurnTracking(), withRetry(),
                       withPersistence({ store, sessionId }));
  const chat = new Conversation({ runner, sessionId, store });

  console.log(chalk.dim(`task: ${task}\n`));
  let streaming = false;
  for await (const event of chat.send(task)) {
    switch (event.type) {
      case "text.delta":
        if (!streaming) { process.stdout.write("  "); streaming = true; }
        process.stdout.write(event.text);
        break;
      case "text.done":
        if (streaming) { process.stdout.write("\n"); streaming = false; }
        break;
      case "tool.done":
        console.log(`  ${chalk.green("✔")} ${chalk.dim(event.toolName)}`);
        break;
      case "tool.error":
        console.log(`  ${chalk.red("✗")} ${chalk.dim(event.toolName)} ${chalk.red(String(event.error))}`);
        break;
      case "error":
        console.error(`  ${chalk.red("✗")} ${event.error.message}`);
        break;
      case "turn.done":
        if (event.usage.totalTokens) console.log(chalk.dim(`  · ${event.usage.totalTokens} tokens`));
        break;
    }
  }

  // C1 — append the full transcript to the flywheel corpus (opt-in).
  const capturePath = process.env.CHP_CAPTURE_TRACES;
  if (capturePath) {
    const msgs = (await store.load(sessionId)) ?? [];
    mkdirSync(dirname(capturePath), { recursive: true });
    appendFileSync(capturePath, JSON.stringify({ messages: msgs, meta: { task, ts: Date.now() } }) + "\n");
    console.log(chalk.dim(`\ncaptured transcript (${msgs.length} messages) → ${capturePath}`));
  }

  await agent.close();
  await closeMCPClients(clients);
  console.log(chalk.dim("\nEvidence: each tool call carried the chp-host MCP session correlation — " +
    "replay it via `chp-host mesh audit`."));
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
